src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,165 @@
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkPositionIndex(index);
+        return getNode(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkPositionIndex(index);
+        Node<T> currentNode = getNode(index);
+        T oldValue = currentNode.element;
+        currentNode.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(getNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> current = head; current != null; current = current.next) {
+            if (object == null && current.element == null
+                    || object != null && object.equals(current.element)) {
+                unlink(current);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void linkNode(T value) {
+        Node<T> currentLastNode = tail;
+        Node<T> newNode = new Node<>(currentLastNode, value, null);
+        tail = newNode;
+        if (currentLastNode != null) {
+            currentLastNode.next = newNode;
+        } else {
+            head = newNode;
+        }
+        size++;
+    }
+
+    private void linkNode(T value, Node<T> nearestNode) {
+        Node<T> previousNode = nearestNode.previous;
+        Node<T> newNode = new Node<>(previousNode, value, nearestNode);
+        nearestNode.previous = newNode;
+        if (previousNode != null) {
+            previousNode.next = newNode;
+        } else {
+            head = newNode;
+        }
+        size++;
+    }
+
+    private Node<T> getNode(int index) {
+        checkElementIndex(index);
+        if (index < (size / 2)) {
+            Node<T> newNode = head;
+            for (int i = 0; i < index; i++) {
+                newNode = newNode.next;
+            }
+            return newNode;
+        }
+        Node<T> newNode = tail;
+        for (int i = size - 1; i > index; i--) {
+            newNode = newNode.previous;
+        }
+        return newNode;
+    }
+
+    private void checkPositionIndex(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index is out of bounds");
+        }
+    }
+
+    private void checkElementIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index is out of bounds");
+        }
+    }
+
+    private T unlink(Node<T> currentNode) {
+        final T element = currentNode.element;
+        final Node<T> next = currentNode.next;
+        final Node<T> previous = currentNode.previous;
```

Comment:
do we need this variables?

Reply:
Yes, because without we can not make removed Node null and it will exist somewhere in our memory. In JDK LinkedList after changing links we should nullify Node that we want to unlink.
+ With this variables code is easier to read.

Reply:
what about garbage collector?